Script for creating a tag with release notes in the message

krystian-hebel opened the pull request:
Example usage:

```bash
./util/dasharo/tag-with-changelog qemu_q35_v0.2.1 qemu_q35_v0.2.2-rc1
```

TBD: this doesn't include changes in the payload, except if explicitly mentioned in changes to config files.

krystian-hebel committed:
util/dasharo/tag-with-changelog: add new file

This is an exact copy of util/release/genrelnotes as of 50e8579bfd
("util/release: Remove makefile.inc references from genrelnotes").
It will be modified for a new role in following commits.

Upstream-Status: Inappropriate [Dasharo downstream]
Change-Id: I32f3d2cb5ad57a52f4b20791f31f2b8ebdfcdd4a
Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

krystian-hebel committed:
util/dasharo/tag-with-changelog: remove unnecessary parts

This includes:
- tool check - it is assumed that git is installed, and other tools are
  no longer required
- lines of code calculations - not required for a changelog
- list of added or removed mainboards, SOCs, CPUs etc.
- number of commits by author and list of new authors

Upstream-Status: Inappropriate [Dasharo downstream]
Change-Id: I17ef764516cd4ca3467e96f39ac0c81b28b8e06a
Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

krystian-hebel committed:
util/dasharo/tag-with-changelog: create a tag with changelog as a message

Apart for new comments for handling the tags, the format of log file
was adjusted so that all informational lines that are not expected in
the final tag description are written as comments. The changelog file
now starts with the instructions for the committer.

This commit also changes the logic checking for clean git tree - the
script no longer requires running from the main branch.

Upstream-Status: Inappropriate [Dasharo downstream]
Change-Id: I9c9b5712fc346a59e9ad0bf7969ee0be1f473a61
Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

krystian-hebel committed:
util/dasharo/tag-with-changelog: dedupe Dasharo boards and build system

Protectli and NovaCustom aren't split to dedicated groups for single
boards, since new platforms are added often for those vendors.

Upstream-Status: Inappropriate [Dasharo downstream]
Change-Id: I2385ec18605f3095580857fdcac0f0e726446794
Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

krystian-hebel commented:
Sometimes two lines are merged into one, I haven't looked as to why yet. It may be related to the fact that I'm currently using the script for non-linear history between two tags (`protectli_vault_jsl_v0.9.3` -> `protectli_vault_jsl_v0.9.4-rc1` + this PR cherry-picked on top). Example:

```
#
##### Intel Braswell (4 commits) #####
- device/device.h: Rename pci_domain_scan_bus- soc/intel/braswell/gpio_support: drop unused get_gpio
- soc/intel/braswell/gpio_support: drop unused get_gpio
- soc/intel/braswell/Kconfig: Set HPET_MIN_TICKS
- soc/intel/braswell: Use {read,write}32p()
```

It seems that there is roughly one such case per group, but it isn't always the first one. The second part is a repetition of another bullet, but is isn't always the immediate one, at least not after sorting.

SergiiDmytruk commented on util/dasharo/tag-with-changelog:
```diff
@@ -45,35 +40,38 @@ else
+	git tag $NEW_GIT_VERSION || exit 1
+
+	# Try to verify that the repo is clean before losing state.
+	if ! git diff-index --quiet --cached HEAD 2>/dev/null || \
+		[ "$(git diff $NEW_GIT_VERSION --shortstat 2>/dev/null | tail -n1)" != "" ]; then
```
Just
```suggestion
	if ! git diff --quiet HEAD; then
```
instead of that upstream horror.

krystian-hebel replied:
It was added as part of https://review.coreboot.org/c/coreboot/+/62440, I suspect it was added to `Handle symbolic links to the git repo`. We probably don't have to deal with this corner case, though, it may be required for some specific downstream build system.

krystian-hebel replied:
https://github.com/Dasharo/coreboot/compare/74d14b8e589dc17fe4381940f7df82be6c8ce81b..4c3f9dc7da460387649522b54bcb68cf2f6a9f31

SergiiDmytruk replied:
> I suspect it was added to `Handle symbolic links to the git repo`

I don't see how, but see 2 new invocations of `realpath` at the bottom which are more likely candidates for handling symlinks.

`git diff-index --quiet --cached HEAD 2>/dev/null` checks for empty index (so staged changes) and `[ "$(git diff $NEW_GIT_VERSION --shortstat 2>/dev/null | tail -n1)" != "" ]` looks like a terrible way for checking for unstaged changes.  Probably a case of unfortunate adjustment while making a release.

SergiiDmytruk commented on util/dasharo/tag-with-changelog:
```diff
@@ -45,35 +40,38 @@ else
+	if ! git diff-index --quiet --cached HEAD 2>/dev/null || \
+		[ "$(git diff $NEW_GIT_VERSION --shortstat 2>/dev/null | tail -n1)" != "" ]; then
+		echo "ERROR: repo is not clean.  Exiting." >&2
+		echo "Deleting temporary tag"
+		git tag -d $NEW_GIT_VERSION
```
With the change above adding a tag and then removing it is unnecessary, just make the tag after the check.

krystian-hebel replied:
https://github.com/Dasharo/coreboot/compare/74d14b8e589dc17fe4381940f7df82be6c8ce81b..4c3f9dc7da460387649522b54bcb68cf2f6a9f31

SergiiDmytruk commented on util/dasharo/tag-with-changelog:
```diff
@@ -396,14 +402,18 @@ for module in $(git submodule --quiet foreach pwd); do
-	mv "$tmpfile" "$MAIN_LOGFILE"
-	rm -f "$LOGFILE"
+# Create signed tag, replace existing, use LOGFILE as message, open editor
+git tag -s -f -F $LOGFILE -e $NEW_GIT_VERSION
```
I'd add quotes around `$LOGFILE` here and below. Version is unlikely to need escaping, but a full path could.  Although, I'd add quotes around `$NEW_GIT_VERSION` too, just because if it's empty or weird in some way, the errors will be more intelligible.

krystian-hebel replied:
https://github.com/Dasharo/coreboot/compare/74d14b8e589dc17fe4381940f7df82be6c8ce81b..4c3f9dc7da460387649522b54bcb68cf2f6a9f31

SergiiDmytruk commented on util/dasharo/tag-with-changelog:
```diff
@@ -0,0 +1,426 @@
+				2>/dev/null; \
+		fi; \
+		if [ -n "$keywords" ]; then \
+			git log --abbrev-commit --pretty=format:"- %s" \
```
```suggestion
			git log --abbrev-commit --pretty=tformat:"- %s" \
```
And above as well.  It makes git add newline character even after printing a single line of output, otherwise some lines get glued as you have already noticed.

krystian-hebel replied:
https://github.com/Dasharo/coreboot/compare/74d14b8e589dc17fe4381940f7df82be6c8ce81b..4c3f9dc7da460387649522b54bcb68cf2f6a9f31

SergiiDmytruk commented on util/dasharo/tag-with-changelog:
```diff
@@ -166,7 +166,7 @@ get_log_submodule() {
 version_ctrl_c() {
 	printf "\n Cleaning up and exiting.\n" >&2
-	git checkout origin/main > /dev/null 2>&1
+	git checkout $NEW_GIT_VERSION > /dev/null 2>&1
```
Then this can also be quoted.

krystian-hebel replied:
https://github.com/Dasharo/coreboot/compare/b20dcc26401c9ec1656c191cc2d9e1c47ab0f743..97105c6159c8ca5a10eccf425310b9f4339ad377

SergiiDmytruk commented on util/dasharo/tag-with-changelog:
```diff
@@ -200,7 +200,7 @@ for module in $(git submodule --quiet foreach pwd); do
 done
 
-git checkout origin/main  > /dev/null 2>&1
+git checkout $NEW_GIT_VERSION  > /dev/null 2>&1
```
As well as this can be quoted.  This makes `shellcheck` happy (there are 2 stylistic notes unrelated to the changes).

krystian-hebel replied:
https://github.com/Dasharo/coreboot/compare/b20dcc26401c9ec1656c191cc2d9e1c47ab0f743..97105c6159c8ca5a10eccf425310b9f4339ad377

SergiiDmytruk commented on util/dasharo/tag-with-changelog:
```diff
@@ -18,62 +18,62 @@ then
 	echo "Old version should be a tag (4.1), a branch (origin/4.1), or a commit id"
-	echo "New version can be a branch (origin/main) a tag (4.2), or a commit id"
-	echo "Logfile can be a new file or an existing file to update"
+	echo "New version must be a name of a new tag, this script fails if it exists"
```
The script no longer fails if the new tag exists because that was a side-effect of creating the tag.  This text or the code needs to change.

krystian-hebel replied:
I've decided to change the code to fail in that case, I think it would be safer than accidentally modifying the tags.

https://github.com/Dasharo/coreboot/compare/b20dcc26401c9ec1656c191cc2d9e1c47ab0f743..97105c6159c8ca5a10eccf425310b9f4339ad377

krystian-hebel committed:
payloads/edk2/Kconfig.dasharo: extract revision and URL from defconfigs

In most cases, edk2 revision is bumped in all Dasharo defconfigs, and
when it isn't, it is usually by mistake. By moving the value to one,
common place, hopefully the number of such mistakes will be smaller.

It will also make it easier to produce the changelog between releases
in semi-automatic way.

Upstream-Status: Inappropriate [Dasharo downstream]
Change-Id: Ic5fe7a9bba347a049d0725b8890dff5f56d32a8a
Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

SergiiDmytruk reviewed (dismissed):
LGTM.  Not merging in case you decide to do further changes here.

krystian-hebel committed:
payloads/Kconfig: introduce DASHARO option

This option is set as an easy indication in defconfigs that the given
platform is released under Dasharo. Use it in workflow for obtaining
the list of defconfigs for Dasharo platforms.

Upstream-Status: Inappropriate [Dasharo downstream]
Change-Id: I3155f16d41742126b6fd769a4d6e12023115e638
Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

krystian-hebel committed:
util/dasharo/tag-with-changelog: list edk2 and submodules commits

Other payloads aren't supported.

Both submodules and edk2 commits are nested one level deeper that the
main log. This is a compromise between enough information (previously
just the number of commit added in a submodule was printed) and clear
grouping between categories.

Changelog for edk2 is limited in history, it starts with the first
commit pointed to by common Kconfig default. For releases made before
this commit, a warning about a possibly incomplete edk2 changelog
will be shown in the comment just above the changelog. The changelog
will be complete only if older tag is made after this commit, in most
cases this will be two releases (or release candidates) from now, for
each platform separately.

Upstream-Status: Inappropriate [Dasharo downstream]
Change-Id: I9db19340b25ba8fe3c8382217e777c9dbe901d04
Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

krystian-hebel commented:
I've added edk2 changelog generation, although it will be working best two releases from now, as there is no easy way of getting common history from multiple defconfigs without somehow parsing tag name and mapping it to config name. Now edk2 revision is common between all Dasharo platforms. In theory it should still be possible to specify different one in config file, but that would once again make things harder to follow.

I don't plan to add anything more in this PR. One possible improvement could be reusing repo in `payloads` instead of downloading a fresh one, but:
- tags aren't created that often so downloading another copy every now and than shouldn't be too annoying
- payloads directories are listed in `.gitignore`, meaning their contents aren't protected by check for clean main repo

SergiiDmytruk commented on util/dasharo/tag-with-changelog:
```diff
@@ -410,6 +422,26 @@ for module in $(git submodule --quiet foreach pwd); do
+	OLD_EDK2_GIT_VERSION="$PREHISTORY"
+fi
+
+git clone "$URL" || version_ctrl_c
```
This should be almost as good as reusing that repository when it exists:
```suggestion
git clone --reference-if-able payloads/external/edk2/workspace/Dasharo/ --bare "$URL" edk2 || version_ctrl_c
```
If that path doesn't exist, the repo gets cloned as usual.  `edk2` at the end is needed because `--bare` results in `edk2.git` name being used by default.

krystian-hebel replied:
https://github.com/Dasharo/coreboot/compare/6116e4b5e35d9173a29ac3fa1e4932ff8b6e60f0..0b42ba7fcb2c81fe05999cd29fd4d93c3cce8555#diff-0ecf7672503aa0c7446284c3c0b00569b847900b9e4e0db39822bf42fb7ba62fR441

It prints a warning when the directory doesn't exist, but since it prints download progress anyway, it may serve as an information as to why the download is required in the first place.

SergiiDmytruk commented on util/dasharo/tag-with-changelog:
```diff
@@ -410,6 +422,26 @@ for module in $(git submodule --quiet foreach pwd); do
+# isn't trivial to differentiate between such case and use as placeholder value.
+PREHISTORY="f0459dff1425c0c02d0736de5737eb23a1b671ec"
+
+if [ -z $OLD_EDK2_GIT_VERSION ]; then
```
```suggestion
if [ -z "$OLD_EDK2_GIT_VERSION" ]; then
```

krystian-hebel replied:
https://github.com/Dasharo/coreboot/compare/6116e4b5e35d9173a29ac3fa1e4932ff8b6e60f0..0b42ba7fcb2c81fe05999cd29fd4d93c3cce8555#diff-0ecf7672503aa0c7446284c3c0b00569b847900b9e4e0db39822bf42fb7ba62fR435

SergiiDmytruk reviewed (commented):
> I've added edk2 changelog generation, although it will be working best two releases from now, as there is no easy way of getting common history from multiple defconfigs without somehow parsing tag name and mapping it to config name.

Name of the defconfig file could be provided on the command-line and `git show "old-revision:configs/$config"` could be used to obtain it, but maybe we don't need to complicate this further given the absence of detailed notes until now.

krystian-hebel committed:
util/dasharo/tag-with-changelog: try to reuse existing edk2 repository

If the script is run in the directory that was used to build coreboot,
chances are that edk2 repository is already cloned in the payloads
subdirectory. It is passed as '--reference-if-able' option, which
reduces the number of objects to be downloaded, even when it is checked
out on older revision. If that path doesn't exist, the repo gets cloned
as usual.

Upstream-Status: Inappropriate [Dasharo downstream]
Change-Id: I11eae8a7d87923a469ac7ba927103395ceb979f4
Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>

krystian-hebel commented:
> Name of the defconfig file could be provided on the command-line and `git show "old-revision:configs/$config"` could be used to obtain it, but maybe we don't need to complicate this further given the absence of detailed notes until now.

This would be just another place to make a mistake. I asked internally about possibility to modify `build.sh` to make config filename and ROM filename consistent (with hope that tag name could be made consistent as well), but I didn't get any answer. The idea would be to add an option to print that data in machine-readable form that could be consumed e.g. by this script.
